src/v/cloud_topics/reconciler/reconciler.cc

```diff
@@ -180,174 +232,353 @@ ss::future<> reconciler::reconciliation_loop() {
     }
 }
 
-ss::future<> reconciler::reconcile() {
-    auto staging_file_fut = co_await ss::coroutine::as_future(
-      _l1_io->create_tmp_file());
-    if (staging_file_fut.failed()) {
-        auto ex = staging_file_fut.get_exception();
-        vlog(lg.error, "Exception creating staging file: {}", ex);
+std::vector<reconciler::attached_partition>
+reconciler::collect_leader_partitions() const {
+    std::vector<attached_partition> leaders;
+    for (const auto& p : _partitions) {
+        if (p.second->partition->is_leader()) {
+            leaders.push_back(p.second);
+        }
     }
-    auto staging_file_result = staging_file_fut.get();
-    if (!staging_file_result.has_value()) {
-        vlog(
-          lg.warn,
-          "Failed to create staging file: {}",
-          static_cast<int>(staging_file_result.error()));
+
+    if (!leaders.empty()) {
+        std::shuffle(
+          leaders.begin(), leaders.end(), random_generators::internal::gen);
+    }
+    return leaders;
+}
+
+ss::future<> reconciler::reconcile() {
+    auto partitions = collect_leader_partitions();
+    if (partitions.empty()) {
+        vlog(lg.debug, "No leader partitions to reconcile");
         co_return;
     }
 
-    auto staging_file = std::move(staging_file_result.value());
-    auto output_stream = co_await staging_file->output_stream();
+    // Begin by creating the set of objects to be built.
+    auto metadata_builder = _metastore->object_builder();
+    absl::flat_hash_map<l1::object_id, std::vector<attached_partition>>
```

Comment:
should we just start off the bat with `chunked_hash_map` and `chunked_vector`? I don't think we care that much about the performance differences here between the two.